crates/core/tedge_api/src/mqtt_topics.rs

```diff
@@ -305,6 +310,75 @@ impl EntityTopicId {
         }
         .map(|parent_id| EntityTopicId(format!("device/{parent_id}//")))
     }
+
+    /// If `self` is a device topic id, return a service topic id under this
+    /// device.
+    ///
+    /// The device topic id must be in a format: "device/DEVICE_NAME//"; if not,
+    /// `None` will be returned.
+    pub fn to_service_topic_id(&self, service_name: &str) -> Option<ServiceTopicId> {
+        if let ["device", device_name, "", ""] = self.0.split('/').collect::<Vec<&str>>()[..] {
+            return Some(ServiceTopicId(EntityTopicId(format!(
+                "device/{device_name}/service/{service_name}"
+            ))));
+        }
+        None
+    }
+
+    /// Returns an array of all segments of this entity topic.
+    pub fn segments(&self) -> [&str; ENTITY_ID_SEGMENTS] {
+        let mut segments = self.0.split('/');
+        let seg1 = segments.next().unwrap();
+        let seg2 = segments.next().unwrap();
+        let seg3 = segments.next().unwrap();
+        let seg4 = segments.next().unwrap();
+        [seg1, seg2, seg3, seg4]
+    }
+
+    pub fn as_str(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+/// Represents an entity topic identifier known to be a service.
+///
+/// It's most often in a format `device/DEVICE_NAME/service/SERVICE_NAME`.
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub struct ServiceTopicId(EntityTopicId);
+
+impl ServiceTopicId {
+    pub fn as_str(&self) -> &str {
+        self.0.as_str()
+    }
+
+    pub fn entity(&self) -> &EntityTopicId {
+        &self.0
+    }
+
+    /// If in a default MQTT scheme, returns a device topic id of this service.
+    pub fn to_device_topic_id(&self) -> Option<DeviceTopicId> {
```

Comment:
We really should avoid these methods that assume a default topic scheme.

Reply:
Yes, this assumption of the default topic scheme, spread everywhere with no good way to indicate it (i.e. using a comment or a custom `.expect()` message, instead of better utilising the type system) leads to some unreliable code, and I think implementing a solution to this should be a priority. I'll look at the current usages and make a ticket.

Reply:
This is why I see more values in the `struct Service` which removes the need for this kind of guessing work.

Reply:
This is why I see more values in the `struct Service` which removes the need for this kind of guessing work.